### ISSUE TITLE
Add support for `or` in ECS querying

### ISSUE THREAD

**Issue by tristanpemble** [bevyengine/bevy#162]
In investigating work for #63 it came up that we cannot do an `or` query right now. The specific query I wanted was:

- Has `A`, `B`, and `C`
- And at least one of those changed

It would be nice to be able to use some form of `or` statement when querying.

**Comment by Emily** [comment]
Summing up some conversation from the Discord on this topic:

- Does `Any<(Component1, Component2)>` make sense?
- `Any` seems most obviously valuable when dealing with `Changed`/`Mutated`/`Added`.
- Perhaps `AnyChanged<(C1, C2)>` (and mutated/added versions) would make more sense?
- It seems a little bit weird to implement `AnyChanged` when a more powerful primitive `Any` could be implemented.
- Tuple approach is nice on `AnyChanged` as it enables many components to be queried at once.
- How would the `Any` API look?
- `Query<(C1, Option<C2>)>` already exists

I'd like to take a stab at implementing the `AnyChanged<(C1, C2, ...)>` approach. It'd result in queries that look something like:

```rust
fn my_system(query: &mut Query<(Component1, AnyChanged<(Component2, Component3))> {
    for (component1, (component2, component3)) in &mut query.iter() {
        // do stuff

        if component2.changed {
            // do more stuff
        }
    }
}
```

I'm not sure if adding the `changed` flag to `Changed` is really all that helpful, but it's easy to do and shouldn't have much downside -- other than being a bit confusing in a context like `Query<Changed<Component1>>` where it'd always be set to `true`.

**Comment by Emily** [comment]
Some more Discord conversation summary:

- adding `changed`/`mutated`/`added` flags to `Changed`/`Mutated`/`Added` can cause some oddities around derefencing
- use a static fn to access the flag, e.g. `if Changed::changed(c1) {`
- also it doesn't really make sense to modify `Changed` and friends, just use new types
- naming new types is hard, but it makes more sense than something like `let (c1_changed, c2_changed) = AnyChanged::changed(r);`

**Comment by BimDav** [comment]
Sorry I did not see your recent comments, Emily. I tried a solution using directly the should_skip methods of two queries to implement Either<Query1, Query2>

**Comment by Emily** [comment]
Even though I really like the tuple approach I outlined, it has a ton of implementation complexity. @BimDav i think your `Either` implementation is the right way to go.

**Comment by cart** [comment]
Closed by #358
